packages/affiliates/scripts/univ3.ts

```diff
@@ -0,0 +1,210 @@
+      nonfungibleTokenPositionDescriptorAddress: "0xeb86f5BE368c3C5e562f7eA1470ACC431d30fB0C",
+      descriptorProxyAddress: "0xB79bDE60fc227217f4EE2102dC93fa1264E33DaB",
+      nonfungibleTokenPositionManagerAddress: "0x865F20efC14A5186bF985aD42c64f5e71C055376",
+      v3MigratorAddress: "0x1988F2e49A72C4D73961C7f4Bb896819d3d2F6a3"
+    }
+  ]
+]);
+const infura = process.env.infura;
+const network = "rinkeby";
+
+function getAddress(network: NetworkName, contractName: string) {
+  const addresses = networks.get(network);
+  assert(addresses[contractName], "invalid contract name: " + contractName);
+  return addresses[contractName];
+}
+
+async function getPoolState(params: { pool: Pool; provider: any }) {
+  const { pool, provider } = params;
+  assert(pool.address, "requires pool address");
+  const contract = new ethers.Contract(pool.address, UniswapV3Pool.abi, provider);
+  const slot0 = await contract.slot0();
+  const liquidity = await contract.liquidity();
+  return {
+    ...slot0,
+    liquidity: liquidity.toString()
+  };
+}
+async function processNftEvents(params: { provider: any; positions: ReturnType<typeof NftPositions> }) {
+  const { provider, positions } = params;
+  const nftHandler = NftEvents({ positions });
+  const contract = new ethers.Contract(
+    getAddress(network, "nonfungibleTokenPositionManagerAddress"),
+    NFTPositionManager.abi,
+    provider
+  );
+  const events = await contract.queryFilter({});
+  await Promise.map(events, nftHandler);
+}
+async function getNftPositionState(params: { provider: any; position: NftPosition }) {
+  const { provider, position } = params;
+  const contract = new ethers.Contract(
+    getAddress(network, "nonfungibleTokenPositionManagerAddress"),
+    NFTPositionManager.abi,
+    provider
+  );
+  return convertValuesToString(await contract.positions(position.tokenId));
+}
+async function processPoolEvents(params: {
+  provider: any;
+  pool: Pool;
+  positions: ReturnType<typeof Positions>;
+  pools: ReturnType<typeof Pools>;
+}) {
+  const { pools, pool, provider, positions } = params;
+  assert(pool.id, "requires pool id");
+  assert(pool.address, "requires pool address");
+  const poolHandler = PoolEvents({ positions, id: pool.id, pools });
+  const contract = new ethers.Contract(pool.address, UniswapV3Pool.abi, provider);
+  const events = await contract.queryFilter({});
+  await Promise.map(events, poolHandler);
+}
+async function getPositionState(params: { position: Position; provider: any; pool: Pool }) {
+  const { position, provider, pool } = params;
+  assert(pool.address, "requires pool address");
+  const contract = new ethers.Contract(pool.address, UniswapV3Pool.abi, provider);
+  return {
+    pool: pool.address,
+    ...convertValuesToString<Position>(await contract.positions(position.id))
+  };
+}
+
+const IsPositionActive = (tick: BigNumberish) => (position: Position) => {
+  assert(position.liquidity, "requires position liquidity");
```

Comment:
Same comment as the one for the Pool type. I think this type should just require these fields, and we can check their existence wherever we ingest.

Reply:
this one unfortunately can't be required, because the event which creates a pool does not have liquidity. So the flow goes like, PoolCreated event -> pool.create -> getPoolState -> pool.update.   The poolCreated event does not have liquidity.

Its possible to fix this with default values, though, which actually might allow for a lot more required keys.